### DISCUSSION TITLE
Selecting a field with a BIGINT data type in MS Access

### DISCUSSION THREAD

**Discussion by Omnitrix2013** [mkleehammer/pyodbc#1307] [Q&A] [answered]
I have created a MS ACCESS table through python using the pyodbc module, like so:
`CREATE TABLE [whole] 
(
    wholeID COUNTER PRIMARY KEY, 
    bin_str VARCHAR(65), 
    twos_comp VARCHAR(65), 
    denary BIGINT, 
    hexadecimal VARCHAR(17)
)`

Note that the BIGINT data type was used for the denary column (the usual INTEGER datatype is just too small for my conversions), however when I try to query this table, I end up getting a strange format back (despite me checking that everything is saved correctly in the MS Access file as can be seen in image).

Shot of output:
![Screenshot 2023-11-25 102216](https://github.com/mkleehammer/pyodbc/assets/77208901/b0870665-b92b-4e30-81ee-77fbca262048)

Shot of another output:
![Screenshot 2023-11-25 102128](https://github.com/mkleehammer/pyodbc/assets/77208901/8c3ef21a-755e-41f4-b941-c65d2c33db5b)

Shot of the MS ACCESS database, which has the correct format for the denary field:
![Screenshot 2023-11-25 102326](https://github.com/mkleehammer/pyodbc/assets/77208901/156b5f1f-4eee-448e-a981-b55895729cb3)

The resultant value for the denary column when I use cursor.fetchall() is in a strange format, for instance: what would usually be "747" becomes "˫\x00\x00\x00", what would usually be "700707" becomes "넣\n\x00\x00". Using the INT datatype instead of the BIGINT fixes it.

This is the kind of simple query that I am running:
`cursor.execute("SELECT * FROM [whole]")`

Including this conversion code didn't change anything:
`def decode_bigint(raw_bytes):
    return int.from_bytes(raw_bytes, byteorder="little")
connection.add_output_converter(pyodbc.SQL_BIGINT, decode_bigint)`

The ODBC trace:
[SQL.LOG](https://github.com/mkleehammer/pyodbc/files/13467202/SQL.LOG)

[info on ODBC trace](https://github.com/mkleehammer/pyodbc/wiki/Troubleshooting-%E2%80%93-Generating-an-ODBC-trace-log)

**Comment by gordthompson** [comment]
Try using `connection.add_output_converter(pyodbc.SQL_WCHAR, decode_bigint)`

**Reply by Omnitrix2013** [reply]
That improved things a lot, I used four test cases and three turned out as expected, so there is still some error going on here, I tried changing the byte order to "big" but that didn't work.
See the image of the output and expected:
![Screenshot 2023-11-26 181842](https://github.com/mkleehammer/pyodbc/assets/77208901/531a9f07-a5f6-41bd-85b3-a4a97cd6d2b7)

**Reply by gordthompson** [reply, marked as answer]
Aha. Try 

```python
def decode_bigint(raw_bytes): return int.from_bytes(raw_bytes, byteorder="little", signed=True)

connection.add_output_converter(pyodbc.SQL_WCHAR, decode_bigint)
```

**Comment by v-chojas** [comment]
Driver bug.

```
pythonw.exe -c  6438-ac88	ENTER SQLGetData 
		HSTMT               0x0000023CBEF104C0
		UWORD                        4 
		SWORD                       -8 <SQL_C_WCHAR>
		PTR                 0x0000023CC173B2A0 
		SQLLEN                  4096
		SQLLEN *            0x000000956E9EE2C8

pythonw.exe -c  6438-ac88	EXIT  SQLGetData  with return code 0 (SQL_SUCCESS)
		HSTMT               0x0000023CBEF104C0
		UWORD                        4 
		SWORD                       -8 <SQL_C_WCHAR>
		PTR                 0x0000023CC173B2A0 [       8] "'\ 0\ 0\ 0"
		SQLLEN                  4096
		SQLLEN *            0x000000956E9EE2C8 (8)
```

pyODBC is requesting the data as a wide string but it appears the driver returns raw 8-byte integer. Of course, even the SQLDescribeCol is wrong as it claims this is an nchar(8) column.
